File: app/src/main/java/com/nagpal/shivam/vtucslab/adapters/ContentAdapter.java
package com.nagpal.shivam.vtucslab.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.nagpal.shivam.vtucslab.models.ContentFile;
import com.nagpal.shivam.vtucslab.models.LabExperiment;
import com.nagpal.shivam.vtucslab.models.LabExperimentSubPart;
import com.nagpal.shivam.vtucslab.R;
import com.nagpal.shivam.vtucslab.utils.StaticMethods;
import com.nagpal.shivam.vtucslab.databinding.LayoutCardSeMspBinding;
import com.nagpal.shivam.vtucslab.databinding.LayoutCardSeSspMfBinding;
import com.nagpal.shivam.vtucslab.databinding.LayoutCardSeSspSfBinding;

import java.util.ArrayList;
import java.util.Arrays;

public class ContentAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int VIEW_TYPE_SSP_SF = 0;
    private static final int VIEW_TYPE_MSP_SF = 1;
    private static final int VIEW_TYPE_SSP_MF = 2;
    private static final int VIEW_TYPE_MSP_MF = 3;
    private static final int VIEW_TYPE_INVALID = -1;
    private Context mContext;
    private ArrayList<LabExperiment> mLabExperimentArrayList;
    private ItemClickHandler mItemClickHandler;

    public ContentAdapter(Context context, ArrayList<LabExperiment> labExperimentArrayList) {
        this.mContext = context;
        this.mLabExperimentArrayList = labExperimentArrayList;
    }

    public void setItemClickHandler(ItemClickHandler itemClickHandler) {
        mItemClickHandler = itemClickHandler;
    }

    public void addAll(ArrayList<LabExperiment> labExperimentArrayList) {
        int i = mLabExperimentArrayList.size();
        mLabExperimentArrayList.addAll(labExperimentArrayList);
        notifyItemRangeInserted(i, labExperimentArrayList.size());
    }

    public void addAll(@NonNull LabExperiment[] labExperiments) {
        int i = mLabExperimentArrayList.size();
        mLabExperimentArrayList.addAll(Arrays.asList(labExperiments));
        notifyItemRangeInserted(i, labExperiments.length);
    }


    public void clear() {
        mLabExperimentArrayList.clear();
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        switch (viewType) {
            case VIEW_TYPE_SSP_SF:
                LayoutCardSeSspSfBinding sspSfBinding = DataBindingUtil.inflate(LayoutInflater.from(mContext), R.layout.layout_card_se_ssp_sf, parent, false);
                return new SspSfViewHolder(sspSfBinding);
            case VIEW_TYPE_MSP_SF:
                LayoutCardSeMspBinding mspSfBinding = DataBindingUtil.inflate(LayoutInflater.from(mContext), R.layout.layout_card_se_msp, parent, false);
                return new MspSfViewHolder(mspSfBinding);
            case VIEW_TYPE_SSP_MF:
                LayoutCardSeSspMfBinding sspMfBinding = DataBindingUtil.inflate(LayoutInflater.from(mContext), R.layout.layout_card_se_ssp_mf, parent, false);
                return new SspMfViewHolder(sspMfBinding);
            case VIEW_TYPE_MSP_MF:
                LayoutCardSeMspBinding seMspBinding = DataBindingUtil.inflate(LayoutInflater.from(mContext), R.layout.layout_card_se_msp, parent, false);
                return new MspMfViewHolder(seMspBinding);
            default:
                return new InvalidViewHolder(new View(mContext));

        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        LabExperiment labExperiment = mLabExperimentArrayList.get(position);
        if (holder instanceof SspSfViewHolder) {
            SspSfViewHolder sspSfViewHolder = (SspSfViewHolder) holder;
            sspSfViewHolder.mBinding.serialOrder.setText(processSerialOrder(labExperiment.getSerialOrder()));
            String[] parts = labExperiment.getLabExperimentSubParts()[0].getContentFiles()[0].getFileName().split("\\.");
            if (parts.length >= 2) {
                sspSfViewHolder.mBinding.programTitle.setText(StaticMethods.formatProgramName(parts[parts.length - 2]));
            }
        } else if (holder instanceof MspSfViewHolder) {
            MspSfViewHolder mspSfViewHolder = (MspSfViewHolder) holder;
            mspSfViewHolder.mBinding.serialOrder.setText(processSerialOrder(labExperiment.getSerialOrder()));
            MultipleSubPartAdapter adapter = new MultipleSubPartAdapter(mContext, labExperiment.getLabExperimentSubParts(), false, mItemClickHandler);
            mspSfViewHolder.mBinding.subPartContainer.setAdapter(adapter);
        } else if (holder instanceof SspMfViewHolder) {
            SspMfViewHolder sspMfViewHolder = (SspMfViewHolder) holder;
            sspMfViewHolder.mBinding.serialOrder.setText(processSerialOrder(labExperiment.getSerialOrder()));
            MultipleFileAdapter adapter = new MultipleFileAdapter(mContext, R.layout.layout_card_single_files_without_sub_parts, labExperiment.getLabExperimentSubParts()[0].getContentFiles(), mItemClickHandler);
            sspMfViewHolder.mBinding.filesContainer.setAdapter(adapter);
        } else if (holder instanceof MspMfViewHolder) {
            MspMfViewHolder mspMfViewHolder = (MspMfViewHolder) holder;
            mspMfViewHolder.mBinding.serialOrder.setText(processSerialOrder(labExperiment.getSerialOrder()));
            MultipleSubPartAdapter adapter = new MultipleSubPartAdapter(mContext, labExperiment.getLabExperimentSubParts(), true, mItemClickHandler);
            mspMfViewHolder.mBinding.subPartContainer.setAdapter(adapter);
        }
    }

    @Override
    public int getItemCount() {
        return mLabExperimentArrayList.size();
    }

    @Override
    public int getItemViewType(int position) {
        LabExperiment experiment = mLabExperimentArrayList.get(position);
        int contentFileLength = 1;
        LabExperimentSubPart[] labExperimentSubParts = experiment.getLabExperimentSubParts();
        for (int i = 0, labExperimentSubPartsLength = labExperimentSubParts.length; i < labExperimentSubPartsLength; i++) {
            LabExperimentSubPart subPart = labExperimentSubParts[i];
            contentFileLength = Math.max(contentFileLength, subPart.getContentFiles().length);
        }
        if (labExperimentSubParts.length == 1 && contentFileLength == 1)
            return VIEW_TYPE_SSP_SF;
        else if (labExperimentSubParts.length > 1 && contentFileLength == 1)
            return VIEW_TYPE_MSP_SF;
        else if (labExperimentSubParts.length == 1 && contentFileLength > 1)
            return VIEW_TYPE_SSP_MF;
        else if (labExperimentSubParts.length > 1 && contentFileLength > 1)
            return VIEW_TYPE_MSP_MF;
        else
            return VIEW_TYPE_INVALID;
    }

    private String processSerialOrder(String order) {
        try {
            int i = Integer.parseInt(order);
            return String.valueOf(i);
        } catch (NumberFormatException e) {
            return order;
        }
    }

    public interface ItemClickHandler {
        void onContentFileClick(ContentFile file);
    }

    class SspSfViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        LayoutCardSeSspSfBinding mBinding;

        SspSfViewHolder(LayoutCardSeSspSfBinding binding) {
            super(binding.getRoot());
            mBinding = binding;
            binding.getRoot().setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            if (mItemClickHandler != null) {
                int position = getAdapterPosition();
                ContentFile contentFile = mLabExperimentArrayList.get(position).getLabExperimentSubParts()[0].getContentFiles()[0];
                mItemClickHandler.onContentFileClick(contentFile);
            }
        }
    }

    class MspSfViewHolder extends RecyclerView.ViewHolder {
        LayoutCardSeMspBinding mBinding;

        MspSfViewHolder(LayoutCardSeMspBinding binding) {
            super(binding.getRoot());
            mBinding = binding;
            mBinding.subPartContainer.setLayoutManager(new LinearLayoutManager(mContext));
            mBinding.subPartContainer.setHasFixedSize(true);
        }
    }

    class SspMfViewHolder extends RecyclerView.ViewHolder {
        LayoutCardSeSspMfBinding mBinding;

        SspMfViewHolder(LayoutCardSeSspMfBinding binding) {
            super(binding.getRoot());
            mBinding = binding;
            mBinding.filesContainer.setLayoutManager(new LinearLayoutManager(mContext));
            mBinding.filesContainer.setHasFixedSize(true);
        }
    }

    class MspMfViewHolder extends RecyclerView.ViewHolder {
        LayoutCardSeMspBinding mBinding;

        MspMfViewHolder(LayoutCardSeMspBinding binding) {
            super(binding.getRoot());
            mBinding = binding;
            mBinding.subPartContainer.setLayoutManager(new LinearLayoutManager(mContext));
            mBinding.subPartContainer.setHasFixedSize(true);
        }
    }

    class InvalidViewHolder extends RecyclerView.ViewHolder {
        InvalidViewHolder(@NonNull View itemView) {
            super(itemView);
        }
    }
}

File: app/src/main/java/com/nagpal/shivam/vtucslab/adapters/ContentAdapter.kt
package com.nagpal.shivam.vtucslab.adapters

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.nagpal.shivam.vtucslab.R
import com.nagpal.shivam.vtucslab.databinding.LayoutCardSeMspBinding
import com.nagpal.shivam.vtucslab.databinding.LayoutCardSeSspMfBinding
import com.nagpal.shivam.vtucslab.databinding.LayoutCardSeSspSfBinding
import com.nagpal.shivam.vtucslab.models.ContentFile
import com.nagpal.shivam.vtucslab.models.LabExperiment
import com.nagpal.shivam.vtucslab.utilities.StaticMethods.formatProgramName
import java.util.*

class ContentAdapter(
    private val context: Context,
    private val labExperimentArrayList: ArrayList<LabExperiment>
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private var itemClickHandler: ItemClickHandler? = null
    fun setItemClickHandler(itemClickHandler: ItemClickHandler) {
        this.itemClickHandler = itemClickHandler
    }

    fun addAll(labExperiments: List<LabExperiment>) {
        val i = labExperimentArrayList.size
        labExperimentArrayList.addAll(labExperiments)
        notifyItemRangeInserted(i, labExperiments.size)
    }

    @SuppressLint("NotifyDataSetChanged")
    fun clear() {
        labExperimentArrayList.clear()
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            VIEW_TYPE_SSP_SF -> {
                val sspSfBinding =
                    DataBindingUtil.inflate<LayoutCardSeSspSfBinding>(
                        LayoutInflater.from(context),
                        R.layout.layout_card_se_ssp_sf,
                        parent,
                        false
                    )
                SspSfViewHolder(sspSfBinding)
            }
            VIEW_TYPE_MSP_SF -> {
                val mspSfBinding =
                    DataBindingUtil.inflate<LayoutCardSeMspBinding>(
                        LayoutInflater.from(context),
                        R.layout.layout_card_se_msp,
                        parent,
                        false
                    )
                MspSfViewHolder(mspSfBinding)
            }
            VIEW_TYPE_SSP_MF -> {
                val sspMfBinding =
                    DataBindingUtil.inflate<LayoutCardSeSspMfBinding>(
                        LayoutInflater.from(context),
                        R.layout.layout_card_se_ssp_mf,
                        parent,
                        false
                    )
                SspMfViewHolder(sspMfBinding)
            }
            VIEW_TYPE_MSP_MF -> {
                val seMspBinding =
                    DataBindingUtil.inflate<LayoutCardSeMspBinding>(
                        LayoutInflater.from(context),
                        R.layout.layout_card_se_msp,
                        parent,
                        false
                    )
                MspMfViewHolder(seMspBinding)
            }
            else -> InvalidViewHolder(View(context))
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val labExperiment = labExperimentArrayList[position]
        val serialOrder = processSerialOrder(labExperiment.serialOrder)
        when (holder) {
            is SspSfViewHolder -> {
                holder.binding.serialOrder.text = serialOrder
                val parts =
                    labExperiment.labExperimentSubParts[0].contentFiles[0].fileName.split("\\.".toRegex())
                        .dropLastWhile { it.isEmpty() }.toTypedArray()
                if (parts.size >= 2) {
                    holder.binding.programTitle.text =
                        formatProgramName(parts[parts.size - 2])
                }
            }
            is MspSfViewHolder -> {
                holder.binding.serialOrder.text = serialOrder
                val adapter = MultipleSubPartAdapter(
                    context,
                    labExperiment.labExperimentSubParts,
                    false,
                    itemClickHandler
                )
                holder.binding.subPartContainer.adapter = adapter
            }
            is SspMfViewHolder -> {
                holder.binding.serialOrder.text = serialOrder
                val adapter = MultipleFileAdapter(
                    context,
                    R.layout.layout_card_single_files_without_sub_parts,
                    labExperiment.labExperimentSubParts[0].contentFiles,
                    itemClickHandler
                )
                holder.binding.filesContainer.adapter = adapter

            }
            is MspMfViewHolder -> {
                holder.binding.serialOrder.text = serialOrder
                val adapter = MultipleSubPartAdapter(
                    context,
                    labExperiment.labExperimentSubParts,
                    true,
                    itemClickHandler
                )
                holder.binding.subPartContainer.adapter = adapter
            }
        }
    }

    override fun getItemCount(): Int {
        return labExperimentArrayList.size
    }

    override fun getItemViewType(position: Int): Int {
        val experiment = labExperimentArrayList[position]
        var contentFileLength = 1
        val labExperimentSubParts = experiment.labExperimentSubParts
        for (subPart in labExperimentSubParts) {
            contentFileLength = contentFileLength.coerceAtLeast(subPart.contentFiles.size)
        }
        val labExperimentSubPartsSize = labExperimentSubParts.size
        return if (labExperimentSubPartsSize == 1 && contentFileLength == 1)
            VIEW_TYPE_SSP_SF
        else if (labExperimentSubPartsSize > 1 && contentFileLength == 1)
            VIEW_TYPE_MSP_SF
        else if (labExperimentSubPartsSize == 1)
            VIEW_TYPE_SSP_MF
        else if (labExperimentSubPartsSize > 1)
            VIEW_TYPE_MSP_MF
        else VIEW_TYPE_INVALID
    }

    private fun processSerialOrder(order: String?): String {
        if (order == null) {
            return "#"
        }
        return try {
            val i = order.toInt()
            i.toString()
        } catch (e: NumberFormatException) {
            order
        }
    }

    interface ItemClickHandler {
        fun onContentFileClick(file: ContentFile)
    }

    internal class InvalidViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
    internal inner class SspSfViewHolder(var binding: LayoutCardSeSspSfBinding) :
        RecyclerView.ViewHolder(
            binding.root
        ), View.OnClickListener {
        init {
            binding.root.setOnClickListener(this)
        }

        override fun onClick(view: View) {
            itemClickHandler?.let {
                val position = adapterPosition
                val contentFile =
                    labExperimentArrayList[position].labExperimentSubParts[0].contentFiles[0]
                it.onContentFileClick(contentFile)
            }
        }
    }

    internal inner class MspSfViewHolder(var binding: LayoutCardSeMspBinding) :
        RecyclerView.ViewHolder(
            binding.root
        ) {
        init {
            binding.subPartContainer.layoutManager = LinearLayoutManager(context)
            binding.subPartContainer.setHasFixedSize(true)
        }
    }

    internal inner class SspMfViewHolder(var binding: LayoutCardSeSspMfBinding) :
        RecyclerView.ViewHolder(
            binding.root
        ) {
        init {
            binding.filesContainer.layoutManager = LinearLayoutManager(context)
            binding.filesContainer.setHasFixedSize(true)
        }
    }

    internal inner class MspMfViewHolder(var binding: LayoutCardSeMspBinding) :
        RecyclerView.ViewHolder(
            binding.root
        ) {
        init {
            binding.subPartContainer.layoutManager = LinearLayoutManager(context)
            binding.subPartContainer.setHasFixedSize(true)
        }
    }

    companion object {
        private const val VIEW_TYPE_SSP_SF = 0
        private const val VIEW_TYPE_MSP_SF = 1
        private const val VIEW_TYPE_SSP_MF = 2
        private const val VIEW_TYPE_MSP_MF = 3
        private const val VIEW_TYPE_INVALID = -1
    }
}
